Userland/DevTools/HackStudio/HackStudioWidget.cpp

```diff
@@ -633,81 +647,133 @@ NonnullRefPtr<GUI::Action> HackStudioWidget::create_new_project_action()
+            tab_widgets.append(child);
             return IterationDecision::Continue;
         });
-        for (size_t i = 0; i < wrappers.size(); ++i) {
-            if (m_current_editor_wrapper.ptr() == &wrappers[i]) {
-                if (i == wrappers.size() - 1)
-                    wrappers[0].editor().set_focus(true);
-                else
-                    wrappers[i + 1].editor().set_focus(true);
+        for (size_t i = 0; i < tab_widgets.size(); ++i) {
+            if (m_current_editor_tab_widget.ptr() == &tab_widgets[i]) {
+                if (i == tab_widgets.size() - 1) {
+                    tab_widgets[0].set_focus(true);
+                    m_current_editor_tab_widget = tab_widgets[0];
+                } else {
+                    tab_widgets[i + 1].set_focus(true);
+                    m_current_editor_tab_widget = tab_widgets[i + 1];
+                }
             }
         }
     });
 }
 
+NonnullRefPtr<GUI::Action> HackStudioWidget::create_switch_to_next_editor_action()
+{
+    return GUI::Action::create("Switch to &Next Editor", { Mod_Ctrl, Key_E }, [this](auto&) {
+        m_current_editor_tab_widget->activate_next_tab();
+    });
+}
+
 NonnullRefPtr<GUI::Action> HackStudioWidget::create_switch_to_previous_editor_action()
 {
     return GUI::Action::create("Switch to &Previous Editor", { Mod_Ctrl | Mod_Shift, Key_E }, [this](auto&) {
-        if (m_all_editor_wrappers.size() <= 1)
-            return;
-        Vector<EditorWrapper&> wrappers;
-        m_editors_splitter->for_each_child_of_type<EditorWrapper>([&wrappers](auto& child) {
-            wrappers.append(child);
-            return IterationDecision::Continue;
-        });
-        for (int i = wrappers.size() - 1; i >= 0; --i) {
-            if (m_current_editor_wrapper.ptr() == &wrappers[i]) {
-                if (i == 0)
-                    wrappers.last().editor().set_focus(true);
-                else
-                    wrappers[i - 1].editor().set_focus(true);
-            }
-        }
+        m_current_editor_tab_widget->activate_previous_tab();
     });
 }
 
 NonnullRefPtr<GUI::Action> HackStudioWidget::create_remove_current_editor_action()
 {
     return GUI::Action::create("&Remove Current Editor", { Mod_Alt | Mod_Shift, Key_E }, Gfx::Bitmap::try_load_from_file("/res/icons/hackstudio/remove-editor.png").release_value_but_fixme_should_propagate_errors(), [this](auto&) {
-        if (m_all_editor_wrappers.size() <= 1)
+        if (m_all_editor_wrappers.size() <= 1 || m_all_editor_tab_widgets.size() <= 1)
             return;
         auto wrapper = m_current_editor_wrapper;
-        m_switch_to_next_editor->activate();
-        m_editors_splitter->remove_child(*wrapper);
+        auto tab_widget = m_current_editor_tab_widget;
+        auto* active_wrapper = tab_widget->active_widget();
+        VERIFY(active_wrapper);
+        tab_widget->on_tab_close_click(*active_wrapper);
+        if (tab_widget->children().is_empty()) {
```

Comment:
This section needs some work. The `on_tab_close_click` function may remove the entire tab widget so perhaps I should move some of this logic to that function?